src/net/client/cache.rs

```diff
@@ -0,0 +1,1160 @@
+                            return response;
+                        }
+                    }
+
+                    let request =
+                        self.upstream.send_request(self.request_msg.clone());
+                    self.state = RequestState::GetResponse(key, request);
+                    continue;
+                }
+                RequestState::GetResponse(key, request) => {
+                    let response = request.get_response().await;
+
+                    let key = key.clone();
+                    let value = Arc::new(Value::new(
+                        &response,
+                        &self.config,
+                        &self.clock,
+                    )?);
+                    self.cache_insert(key, value).await;
+
+                    return response;
+                }
+                RequestState::GetResponseNoCache(request) => {
+                    return request.get_response().await;
+                }
+            }
+        }
+    }
+
+    /// Try to find an cache entry for the key.
+    async fn cache_lookup(
+        &self,
+        key: &Key,
+    ) -> Result<Option<Arc<Value<C>>>, Error> {
+        // There are 4 flags that may affect the response to a query.
+        // In some cases the response to one value of a flag could be
+        // used for the ohter value.
+        // This function takes all 4 flags. First we fix the CD flag.
+        // This flag has to be used as is. Next we have the request to
+        // a function that looks at RD, DO, and AD.
+        self.cache_lookup_rd_do_ad(key).await
+    }
+
+    /// Try to find an cache entry for the key taking into account the
+    /// RD, DO, and AD flags.
+    async fn cache_lookup_rd_do_ad(
+        &self,
+        key: &Key,
+    ) -> Result<Option<Arc<Value<C>>>, Error> {
+        // For RD=1 we can only use responses to queries with RD set.
+        // For RD=0, first try with RD=0 and then try with RD=1. If
+        // RD=1 has an answer, store it as an answer for RD=0.
+        let opt_value = self.cache_lookup_do_ad(key).await?;
+        if opt_value.is_some() || key.rd {
+            return Ok(opt_value);
+        }
+
+        // Look if there is something with RD=1. We can use the
+        // response unmodified.
+        let mut alt_key = key.clone();
+        alt_key.rd = true;
+        let opt_value = self.cache_lookup_do_ad(&alt_key).await?;
+        if let Some(value) = &opt_value {
+            match &value.response {
+                Err(_) => {
+                    // Just insert the error
+                    self.cache_insert(key.clone(), value.clone()).await;
+                }
+                Ok(msg) => {
+                    let msg = clear_rd(msg)?;
+                    let value =
+                        Arc::new(Value::<C>::new_from_value_and_response(
+                            value.clone(),
+                            Ok(msg),
+                            &self.config,
```

Comment:
I'm a bit concerned about the `clear_xxx()` functions as they don't actually clear anything in the passed value, they clone it, clear the flag in the clone, and return the clone, and this is not obvious from the function name. I would find it more like canonical Rust if the conversions were explicit and instead of adding one `clear_xxx()` fn per flag being cleared just use the existing flag setting functions, e.g.:

```
let mut msg: Message::<BytesMut> = msg.into();
msg.header_mut().set_rd(false);
let value =
    Arc::new(Value::<C>::new_from_value_and_response(
        value.clone(),
        Ok(msg.into()),
        &self.config,
    )?);
```

With these defined somewhere:

```
impl From<&Message<Bytes>> for Message<BytesMut> {
    fn from(msg: &Message<Bytes>) -> Self {
        // We started from a Message which was already checked so this shouldn't fail.
        Message::from_octets(msg.as_slice().into()).unwrap()
    }
}

impl From<Message<BytesMut>> for Message<Bytes> {
    fn from(msg: Message<BytesMut>) -> Self {
        // We started from a Message which was already checked so this shouldn't fail.
        Message::from_octets(msg.into_octets().freeze()).unwrap()
    }
}
```

Reply:
I don't see how your version is better. Mine is a clear function, it take an input message and returns a modified one. If there is a better way to name those functions then of course the names can be changed.

Your version seems to move a lot of stuff to the caller and hides the clones in the '.into' calls making it less clear what is going on.

Reply:
Leave it then, I find it harder to read your code, you find it harder to read mine ;-)